tosfs/core.py

```diff
@@ -203,6 +203,59 @@ def info(
 
         return self._try_dir_info(bucket, key, path, fullpath)
 
+    def rmdir(self, path: str) -> None:
+        """Remove a directory if it is empty.
+
+        Parameters
+        ----------
+        path : str
+            The path of the directory to remove. The path should be in the format
+            `tos://bucket_name/directory_path`.
+
+        Raises
+        ------
+        FileNotFoundError
+            If the directory does not exist.
+        NotADirectoryError
+            If the path is not a directory.
+        TosfsError
+            If the directory is not empty,
+             or the path is a bucket.
+
+        Examples
+        --------
+        >>> fs = TosFileSystem()
+        >>> fs.rmdir("tos://mybucket/mydir/")
+
+        """
+        path = self._strip_protocol(path).rstrip("/") + "/"
+        bucket, key, _ = self._split_path(path)
+        if not key:
+            logger.error("Cannot remove a bucket using rmdir api.")
+            raise TosfsError("Cannot remove a bucket using rmdir api.")
+
+        if not self.exists(path):
+            raise FileNotFoundError(f"Directory {path} not found.")
+
+        if not self.isdir(path):
+            raise NotADirectoryError(f"{path} is not a directory.")
+
+        if len(self.ls(path, refresh=True, detail=False)) > 0:
+            raise TosfsError(f"Directory {path} is not empty.")
+
+        try:
+            self.tos_client.delete_object(bucket, key.rstrip("/") + "/")
```

Comment:
I think we will need to introduce a delegate tos client, so that we can handle the TOS error code retry actions in the delegation layer.  Assume that if the TOS server return an error code, then in theory the TOS client should retry with the backoff algorithm, so that it won't interrupt the client's process...

Reply:
Yes, I also think robustness and stability are very important. We will introduce retry strategies and provide more robust responses for specific exception status codes, even performance tests. I temporarily regard these as non-functional improvements. I think currently, the top priority is to make some basic APIs work to ensure that CI can operate normally without relying on the native SDK of tos. The work on stability will be specifically re-examined in the near future. What do you think?

Reply:
I think it's okay, but I hope we can file an issue to track this thing.

Reply:
The issue is here: #25